scipy/stats/_survival.py

```diff
@@ -374,3 +391,138 @@ def _ecdf_right_censored(sample):
+    a different type of tumor. The week to death of 51 adults was recorded.
+
+    >>> from scipy import stats
+    >>> x = stats.CensoredData(
+    ...     uncensored=[6, 13, 21, 30, 37, 38, 49, 50,
+    ...                 63, 79, 86, 98, 202, 219],
+    ...     right=[31, 47, 80, 82, 82, 149]
+    ... )
+    >>> y = stats.CensoredData(
+    ...     uncensored=[10, 10, 12, 13, 14, 15, 16, 17, 18, 20, 24, 24,
+    ...                 25, 28,30, 33, 35, 37, 40, 40, 46, 48, 76, 81,
+    ...                 82, 91, 112, 181],
+    ...     right=[34, 40, 70]
+    ... )
+
+    We can calculate and visualize the empirical survival functions
+    of both groups as follows.
+
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> ax = plt.subplot()
+    >>> ecdf_x = stats.ecdf(x)
+    >>> ax.step(np.insert(ecdf_x.x, 0, 0),
+    ...         np.insert(ecdf_x.sf.points, 0, 1),
+    ...         where='post', label='Astrocytoma')
+    >>> ecdf_y = stats.ecdf(y)
+    >>> ax.step(np.insert(ecdf_y.x, 0, 0),
+    ...         np.insert(ecdf_y.sf.points, 0, 1),
+    ...         ls='-.', where='post', label='Glioblastoma')
+    >>> ax.set_xlabel('Time to death (weeks)')
+    >>> ax.set_ylabel('Empirical SF')
+    >>> plt.legend()
+    >>> plt.show()
+
+    Based on a visual inspection of the survival functions, we see a difference
+    but cannot conclude about its significance.
+
+    Next, we will use the logrank test to assess wether the difference
+    between the two survival functions is significant.
+
+    >>> res = stats.log_rank(x=x, y=y)
+    >>> res.statistic
+    7.49659416854
+    >>> res.pvalue
+    0.006181578637
+
+    Using a significance level of 5%, we would reject the null hypothesis in
+    favor of the alternative hypothesis: "there is a difference between the
+    two survival functions".
+
+    """
+    x = _iv_CensoredData(sample=x, param_name='x')
+    y = _iv_CensoredData(sample=y, param_name='y')
+
+    xy = CensoredData(
+        uncensored=np.concatenate((x._uncensored, y._uncensored)),
+        right=np.concatenate((x._right, y._right))
+    )
+
+    res = ecdf(xy)
+    idx = res.sf._d.astype(bool)  # events were observed (not censored)
+    times_xy = res.x[idx]
+    at_risk_xy = res.sf._n[idx]
+    deaths_xy = res.sf._d[idx]
+
+    n_died_x = x._uncensored.size
+
+    res_x = ecdf(x)
+    i = np.searchsorted(res_x.x, times_xy)  # or use `interpolate_1d`
+    at_risk_x = np.append(res_x.sf._n, 0)[i]  # 0 at risk after last x time
+    at_risk_y = at_risk_xy - at_risk_x
+
+    sum_exp_deaths_x = np.sum(at_risk_x * (deaths_xy/at_risk_xy))
+
+    # equivalent: formulation with the mean
+    # statistic = (
+    #     (n_died_x - sum_exp_deaths_x)**2/sum_exp_deaths_x
+    #     + (n_died_y - sum_exp_deaths_y)**2/sum_exp_deaths_y
+    # )
+    # warning occurs when multiple deaths at same time
+    with np.errstate(invalid='ignore'):
+        sum_var = np.nansum(
```

Comment:
I'd like to track down some mention of why it's appropriate to ignore NaNs here.

Reply:
I took a closer look and this happens when `at_risk_xy  == 1`. We can't have more deaths than those at risk, and we are only interested in times at which there is at least one death, so `deaths_xy == at_risk_xy == 1`. The `at_risk_xy - deaths_xy == 0` in the numerator and the `at_risk_xy - 1 == 0` in the denominator generate a NaN. But assuming it is OK to take the limit, we would find that these terms cancel. The rest of the denominator is 1, and one of `at_risk_x` or `at_risk_y` must be zero, so the term would be zero and doesn't need to be considered.

Another way to express this that would avoid the warnings would be:
```
    num = at_risk_x*at_risk_y*deaths_xy*(at_risk_xy - deaths_xy)
    den = at_risk_xy**2*(at_risk_xy - 1)
    i = at_risk_xy > 1  # when `at_risk_xy == 1`, variance is 0
    sum_var = np.sum(num[i]/den[i])
```

Reply:
I am fine with both. If you don't like having a warning+nansum+explanation then yes this seems better.

Reply:
I'm torn. It's fine to leave it if you prefer. I was mostly sensitive to the NaN thing because of all the NaN discrepancies between different software for Kaplan -Meier confidence intervals.

Reply:
I usually find it cleaner without a clause like `with np.errstate`. That clause could potentially ignore other errors that one dies not think of at first.

Reply:
For the nansum and error, I think with the comment and our analysis it's ok.

I decided to go for returning the `statistic` and make a note in the docstrings about what we do. Doing so I also added alternative (which I could remove for now.) Not sure about tests there though.

Reply:
@lorentzenchr I tend to agree, and I also prefer to avoid nested indentations when the same thing can be achieved without indentation. But I'll mark this as resolved since it is a matter of taste.

Reply:
Please do feel free to raise a new comment about it, though, @lorentzenchr.

Reply:
To be clear, I don't really have strong opinions on that part (the code you proposed is readable to me and close to the formula) so feel free to change this as you push the other changes you mentioned 😃